### PR TITLE
RFC implementation

### PR DESCRIPTION
**Pull request by gabriel-ss**
We are starting to integrate Julia in the company I work, and we needed a tool akin to python's [coverage report](https://coverage.readthedocs.io/en/6.4.4/cmd.html#cmd-report). LocalCoverage provided most of the functionality that we wanted, but lacked the ability to display uncovered lines, so I implemented it. We thought that the functionality might be useful for other people and decided to open a PR, with the new feature:
image

When I went to open the PR I found the #18 and decided to work on it. This PR is my implementation proposal; I tried to respect use case described in #28 and the QoL features proposed in https://github.com/JuliaCI/LocalCoverage.jl/issues/18#issuecomment-818708330, while trying to stay close to what was described. I also removed the XML generation function since it was just a wrapper around lcov_cobertura and wasn't included in the RFC, but I can re-add it if that wasn't the intention. Lastly, I added some convenience features to facilitate use in shell scripts, e.g.

```bash
julia --project -e'using LocalCoverage; report(target_coverage=90)'
```

If there is interest in this implementation, but it somehow deviates too much from the current vision for the package, please let me know what I can do to accommodate it.

### CODE REVIEW

File: src/LocalCoverage.jl [hunk 1]
```diff
@@ -4,199 +4,224 @@ using CoverageTools
 using DocStringExtensions
 using Printf
 using PrettyTables
+using DefaultApplication
 import Pkg
 
-export generate_coverage, open_coverage, clean_coverage, coverage_summary
+export generate_coverage, report, html_coverage
```
Comment (tpapp) [inline src/LocalCoverage.jl:10]: `report` may be too general to export unconditionally, what about `report_coverage`?

File: src/LocalCoverage.jl [hunk 2]
```diff
@@ -4,199 +4,224 @@ using CoverageTools
...
 "Directory for coverage results."
 const COVDIR = "coverage"
 
 "Coverage tracefile."
 const LCOVINFO = "lcov.info"
 
-const PYTHON = get!(ENV, "PYTHON", isnothing(Sys.which("python3")) ? "python" : "python3")
 
+struct FileCoverageSummary
+    filename::String
```
Comment (tpapp) [inline src/LocalCoverage.jl:20]: Please document fields.

File: src/LocalCoverage.jl [hunk 3]
```diff
@@ -4,199 +4,224 @@ using CoverageTools
...
+    lines_hit::Int
+    lines_tracked::Int
+    coverage::Float64
+    coverage_gaps::Vector{UnitRange{Int}}
+end
+
+struct PackageCoverage
+    package_dir::String
```
Comment (tpapp) [inline src/LocalCoverage.jl:28]: Please document fields.

File: src/LocalCoverage.jl [hunk 4]
```diff
@@ -4,199 +4,224 @@ using CoverageTools
...
 
 "Directory for coverage results."
 const COVDIR = "coverage"
 
 "Coverage tracefile."
 const LCOVINFO = "lcov.info"
 
-const PYTHON = get!(ENV, "PYTHON", isnothing(Sys.which("python3")) ? "python" : "python3")
 
+struct FileCoverageSummary
```
Comment (tpapp) [inline src/LocalCoverage.jl:19]:
```suggestion
Base.@kwdef struct FileCoverageSummary
```

File: src/LocalCoverage.jl [hunk 5]
```diff
@@ -4,199 +4,224 @@ using CoverageTools
...
+    filename::String
+    lines_hit::Int
+    lines_tracked::Int
+    coverage::Float64
+    coverage_gaps::Vector{UnitRange{Int}}
+end
+
+struct PackageCoverage
```
Comment (tpapp) [inline src/LocalCoverage.jl:27]:
```suggestion
Base.@kwdef struct PackageCoverage
```

File: src/LocalCoverage.jl [hunk 6]
```diff
@@ -4,199 +4,224 @@ using CoverageTools
...
+        FileCoverageSummary(file.filename, hit, tracked, 100 * hit / tracked, gaps)
     end
 
-    tab[n+1, :] .= ("TOTAL", (total_hit, total_tracked),
-                    100 * total_hit / total_tracked)
-    tab
+    total_hit = sum(getfield.(coverage_list, :lines_hit))
+    total_tracked = sum(getfield.(coverage_list, :lines_tracked))
+
+    PackageCoverage(
```
Comment (tpapp) [inline src/LocalCoverage.jl:84]: I would suggest using a keyword constructor. (post-1.6 non-redundant syntax is fine).

**Review by tpapp — CHANGES_REQUESTED** [review]

**Comment by tpapp** [comment]
I only had time for a superficial review now, but it looks great, thanks for the very clean code. Feel free to depend on Julia 1.6 and bump versions in CI etc accordingly.

**Comment by gabriel-ss** [comment]
Done! Let me know if more changes are needed.

File: src/LocalCoverage.jl [hunk 7]
```diff
@@ -37,7 +37,6 @@ $(SIGNATURES)
 
 Get the root directory of a package.
 """
-pkgdir(m::Module) = joinpath(dirname(pathof(m)), "..")
```
Comment (tpapp) [inline src/LocalCoverage.jl]: Why was this removed? Using modules should be possible (if package if already loaded).

File: src/LocalCoverage.jl [hunk 8]
```diff
@@ -37,8 +37,7 @@ $(SIGNATURES)
 
 Get the root directory of a package.
 """
-pkgdir(pkgstr::AbstractString) =
-    joinpath(dirname(Base.locate_package(Base.PkgId(pkgstr))), "..")
+pkgdir(pkgstr::AbstractString) = abspath(joinpath(dirname(Base.find_package(pkgstr)), ".."))
```
Comment (tpapp) [inline src/LocalCoverage.jl:64]: I am fine with this for now but it is very likely to be fragile, cf https://discourse.julialang.org/t/path-of-this-package/16401/21.

File: src/LocalCoverage.jl [hunk 9]
```diff
@@ -4,199 +4,246 @@ using CoverageTools
...
+"""
+$(SIGNATURES)
 
-        tab[i,:] .= (f.filename, (hit, tracked),
-                     100 * hit / tracked)
+Evaluate the ranges of lines without coverage.
+"""
+function find_gaps(coverage)
+    i = 1
+    last_line = length(coverage)
```
Comment (tpapp) [inline src/LocalCoverage.jl:73]:
```suggestion
    i, last_line = firstindex(coverage), lastindex(coverage)
```
(generalized indexing)

File: src/LocalCoverage.jl [hunk 10]
```diff
@@ -4,199 +4,246 @@ using CoverageTools
...
-        tracked = count(x->!isnothing(x),        f.coverage)
-
-        total_hit     += hit
-        total_tracked += tracked
+"""
+$(SIGNATURES)
 
-        tab[i,:] .= (f.filename, (hit, tracked),
-                     100 * hit / tracked)
+Evaluate the ranges of lines without coverage.
```
Comment (tpapp) [inline src/LocalCoverage.jl:69]:
```suggestion
Evaluate the ranges of lines without coverage.

`coverage` is the vector of coverage counts for each line.
```

File: src/LocalCoverage.jl [hunk 11]
```diff
@@ -4,199 +4,246 @@ using CoverageTools
...
+        eval_coverage_metrics(coverage, package_dir)
+    end
 end
 
-"""
-$(SIGNATURES)
 
-Generate a coverage report for package `pkg`.
+format_gap(gap) = length(gap) == 1 ? "$(first(gap))" : "$(first(gap)) - $(last(gap))"
+format_line(summary) = hcat(
```
Comment (tpapp) [inline src/LocalCoverage.jl:150]: please use `function format_line ... end` for long functions.

File: src/LocalCoverage.jl [hunk 12]
```diff
@@ -4,199 +4,246 @@ using CoverageTools
...
+
+Generate, and optionally open, the HTML coverage summary in a browser for `pkg`
+inside `dir`.
+
+See [`generate_coverage`](@ref).
+"""
+function html_coverage(coverage::PackageCoverage; open = false, dir = tempdir())
+    cd(coverage.package_dir) do
+        branch = try
+            strip(read(`git rev-parse --abbrev-ref HEAD`, String))
```
Comment (tpapp) [inline src/LocalCoverage.jl:197]: Any chance we could use [LibGit2](https://docs.julialang.org/en/v1/stdlib/LibGit2/) for this query?

File: test/runtests.jl [hunk 13]
```diff
@@ -9,12 +13,21 @@ if !isfile(lockfile)
...
+        html_coverage("LocalCoverage", dir = dir)
+        @test isfile(joinpath(dir, "index.html"))
+    end
+
     rm(lockfile)
     @test isfile(tracefile)
-    clean_coverage("LocalCoverage", rm_directory=true)
-    @test !isdir(covdir)
+    rm(covdir, recursive = true)
 end
```
Comment (tpapp) [inline test/runtests.jl:33]: A trivial test for `find_gaps` would be nice.

**Review by tpapp — CHANGES_REQUESTED** [review]
Thanks for the updates, I had a chance to read through this now. It is very nice, just asked for a couple of minor things.

The path handling mechanism is still hacky, but that is outside the scope of this PR, so I opened #31.

File: src/LocalCoverage.jl (hunk 7 above)
Comment (gabriel-ss) [reply src/LocalCoverage.jl]: Pkg.test doesn't have an overload for modules:
```julia
julia> Pkg.test(LocalCoverage)
ERROR: MethodError: no method matching test(::Module)
Closest candidates are:
  test(; name, uuid, version, url, rev, path, mode, subdir, kwargs...) at /usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:162
  test(::Union{AbstractString, Pkg.Types.PackageSpec}; kwargs...) at /usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:143
  test(::Vector{<:AbstractString}; kwargs...) at /usr/share/julia/stdlib/v1.8/Pkg/src/API.jl:144
```

File: src/LocalCoverage.jl (hunk 8 above)
Comment (gabriel-ss) [reply src/LocalCoverage.jl:64]: I don't think there is a public API for finding package names currently, but yeah, I don't like this solution either.

File: src/LocalCoverage.jl (hunk 9 above)
Comment (gabriel-ss) [reply src/LocalCoverage.jl]: Nice one

File: src/LocalCoverage.jl (hunk 12 above)
Comment (gabriel-ss) [reply src/LocalCoverage.jl]: Sure! c1bc763

File: test/runtests.jl (hunk 13 above)
Comment (gabriel-ss) [reply test/runtests.jl:36]: Added in 8012b7d

File: src/LocalCoverage.jl (hunk 11 above)
Comment (gabriel-ss) [reply src/LocalCoverage.jl]: Refactored format_line in c1bc763. I think it was the longest one using the shorter syntax, but let me know if there are any other functions that you'd like to be refactored as well.

**Comment by gabriel-ss** [comment]
I think that covers everything, let me know if you would prefer 007c2c6 to be reverted.

**Pull request merged**